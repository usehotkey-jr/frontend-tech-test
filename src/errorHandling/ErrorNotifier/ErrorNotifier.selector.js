// @flow

import {createSelector} from "reselect";

export const selectErrorNotifierProps = createSelector(
    state => state.errorHandling,
    errorHandling => ({errors: errorHandling.all})
);
