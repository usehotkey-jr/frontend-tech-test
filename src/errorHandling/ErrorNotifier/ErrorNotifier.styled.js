// @flow

import styled from "styled-components";
import {size} from "../../theme/size";
import {palette} from "../../theme/palette";

const {medium, small, large} = size;

export const ContainerEN = styled.div`
  position: fixed;
  
  right: ${medium.spacing};
  top: ${medium.spacing};
`;

export const NotificationItem = styled.div`
  max-width: ${large.block};
  padding: ${small.spacing};
  margin-bottom: ${medium.spacing};
  
  border-radius: ${medium.radius};
  color: ${palette.fg.accent};
  background-color: ${palette.badge.danger};
  
  word-break: break-word;
`;
