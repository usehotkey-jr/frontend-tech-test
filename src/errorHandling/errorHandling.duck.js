// @flow

import type {Action, ActionCreator} from "../_helpers/types/redux";
import {createActionCreatorFactory} from "../_helpers/redux/createActionCreatorFactory";
import {isActionOfType} from "../_helpers/redux/isActionOfType";
import type {Response} from "../api/request";

const actionCreatorFactory = createActionCreatorFactory("ERROR_HANDLING");

export type ErrorHandling = {
    stack: mixed[];
    lastApi?: Response<mixed>;
};

const errorHandlingInitialState = {
    stack: [],
    lastApi: undefined,
};


/**
 * Reducer for errorHandling
 */
export function errorHandlingReducer (state: ErrorHandling = errorHandlingInitialState, action: Action<any>) {
    if (isActionOfType(action, handleApi)) {
        const error = action.payload;

        return {
            stack: [...state.stack.slice(0, 9), error],
            lastApi: error,
        };
    }

    return state;
}

const handleApi: ActionCreator<Response<mixed>> = actionCreatorFactory("HANDLE_API");

export const errorHandlingActions = {
    handleApi,
};
