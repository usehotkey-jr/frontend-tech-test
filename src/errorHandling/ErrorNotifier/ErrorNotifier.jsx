// @flow

import React from "react";
import {connect} from "react-redux";
import {ContainerEN, NotificationItem} from "./ErrorNotifier.styled";
import {selectErrorNotifierProps} from "./ErrorNotifier.selector";
import type {Notification} from "../errorHandling.duck";
import {errorHandlingActions} from "../errorHandling.duck";

type ErrorNotifierStateProps = {
    errors: Notification[];
};

type ErrorNotifierDispatchProps = {
    clear: typeof errorHandlingActions.clear;
};

type ErrorNotifierProps = ErrorNotifierStateProps & ErrorNotifierDispatchProps;

const NOTIFICATION_LIVE_TIME = 5000;

export class ErrorNotifierComponent extends React.PureComponent<ErrorNotifierProps> {
    componentWillReceiveProps (nextProps: ErrorNotifierStateProps) {
        nextProps.errors.forEach(nextError => {
            if (this.props.errors.indexOf(nextError) === -1) {
                setTimeout(() => this.props.clear(nextError), NOTIFICATION_LIVE_TIME);
            }
        });
    }

    render () {
        const errorHtml = this.props.errors
            .map(error => (<NotificationItem key={error.id}>{error.message}</NotificationItem>));

        return (
            <ContainerEN>
                {errorHtml}
            </ContainerEN>
        );
    }
}

export const ErrorNotifier = connect(selectErrorNotifierProps, {
    clear: errorHandlingActions.clear,
})(ErrorNotifierComponent);
