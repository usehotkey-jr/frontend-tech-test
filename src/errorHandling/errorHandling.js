// @flow

import {errorHandlingActions} from "./errorHandling.duck";
import type {Response} from "../api/request";
import {getStore} from "../app/store";

/**
 * Handler of api errors.
 */
export function handleApiError (response: any, error: Response<mixed>) {
    const handledError = {
        message: `Request to ${response.url} failed with status: ${response.status}.
        Error message: ${error.message}`,
        id: Math.random(),
    };

    getStore().dispatch(errorHandlingActions.handle(handledError));
}

window.addEventListener("unhandledrejection", event => {
    event.preventDefault();
    console.warn("Unhandled promise rejection:", event); // eslint-disable-line
});
