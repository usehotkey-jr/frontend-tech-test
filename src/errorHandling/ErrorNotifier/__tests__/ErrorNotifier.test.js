import {createTestSnapshot} from "../../../_helpers/jest/createTestSnapshot.ignore";
import {ErrorNotifierComponent} from "../ErrorNotifier";
import {errorsMock} from "../../__mocks__/errorHandling.mock";

describe("ErrorNotifier", () => {
    test("should correctly render", () => {
        const props = {
            errors: errorsMock,
        };
        createTestSnapshot.all(ErrorNotifierComponent, props);
    });
});
