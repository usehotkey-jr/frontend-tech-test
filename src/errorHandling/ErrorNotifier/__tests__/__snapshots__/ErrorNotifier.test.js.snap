// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorNotifier should correctly render 1`] = `
<styled.div>
  <styled.div
    key="1"
  >
    123
  </styled.div>
  <styled.div
    key="2"
  >
    234
  </styled.div>
</styled.div>
`;

exports[`ErrorNotifier should correctly render 2`] = `
<div
  class="sc-bdVaJa bBbzvV"
>
  <div
    class="sc-bwzfXH dNrWeL"
  >
    123
  </div>
  <div
    class="sc-bwzfXH dNrWeL"
  >
    234
  </div>
</div>
`;

exports[`ErrorNotifier should correctly render 3`] = `
<ErrorNotifierComponent
  errors={
    Array [
      Object {
        "id": 1,
        "message": "123",
      },
      Object {
        "id": 2,
        "message": "234",
      },
    ]
  }
>
  <styled.div>
    <div
      className="sc-bdVaJa bBbzvV"
    >
      <styled.div
        key="1"
      >
        <div
          className="sc-bwzfXH dNrWeL"
        >
          123
        </div>
      </styled.div>
      <styled.div
        key="2"
      >
        <div
          className="sc-bwzfXH dNrWeL"
        >
          234
        </div>
      </styled.div>
    </div>
  </styled.div>
</ErrorNotifierComponent>
`;
